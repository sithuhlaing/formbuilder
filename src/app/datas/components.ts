export const components = {
  input_components: {
    icon: "✏️",
    label: "Input Components",
    components: [
      {
        type: "text_input",
        icon: "📝",
        label: "Text Input",
        description: "Single line text field",
        properties: {
          label: "Text Input",
          placeholder: "Enter text",
          required: false,
        },
      },
      {
        type: "email_input",
        icon: "📧",
        label: "Email Input",
        description: "Email address field with validation",
        properties: {
          label: "Email Input",
          placeholder: "Enter email",
          required: true,
        },
      },
      {
        type: "password_input",
        icon: "🔒",
        label: "Password Input",
        description: "Password field with masking",
        properties: {
          label: "Password Input",
          placeholder: "Enter password",
          required: true,
        },
      },
      {
        type: "number_input",
        icon: "🔢",
        label: "Number Input",
        description: "Numeric input with validation",
        properties: {
          label: "Number Input",
          placeholder: "Enter number",
          required: false,
        },
      },
      {
        type: "textarea",
        icon: "📄",
        label: "Text Area",
        description: "Multi-line text input",
        properties: {
          label: "Text Area",
          placeholder: "Enter long text",
          required: false,
        },
      },
      {
        type: "rich_text",
        icon: "✍️",
        label: "Rich Text Editor",
        description: "WYSIWYG text editor",
        properties: {
          label: "Rich Text Editor",
          content: "Initial content",
        },
      },
      {
        type: "patient_id",
        icon: "🆔",
        label: "Patient ID",
        description: "Standardised patient identifier with validation",
        properties: {
          label: "Patient ID",
          placeholder: "e.g. PAT-123456",
          required: true,
          validation: [
            { type: "required", message: "Patient ID is required" },
            {
              type: "pattern",
              pattern: "^PAT-[0-9]{6}$",
              message: "Use format PAT-123456",
            },
          ],
        },
      },
      {
        type: "medical_record_number",
        icon: "📁",
        label: "Medical Record Number",
        description: "Unique MRN capture with format guardrails",
        properties: {
          label: "Medical Record Number",
          placeholder: "e.g. MRN-00000001",
          required: true,
          validation: [
            { type: "required", message: "Medical record number is required" },
            {
              type: "pattern",
              pattern: "^MRN-[0-9]{8}$",
              message: "Use format MRN-00000000",
            },
          ],
        },
      },
      {
        type: "insurance_member_number",
        icon: "💳",
        label: "Insurance Member Number",
        description: "Capture insurance membership identifiers",
        properties: {
          label: "Insurance Member Number",
          placeholder: "e.g. HMO-1234-5678",
          required: false,
          validation: [
            {
              type: "pattern",
              pattern: "^[A-Z]{2,3}-[0-9]{4}-[0-9]{4}$",
              message: "Use format HMO-1234-5678",
            },
          ],
        },
      },
      {
        type: "nhs_number",
        icon: "🏥",
        label: "NHS Number",
        description: "UK NHS compliant patient identifier",
        properties: {
          label: "NHS Number",
          placeholder: "000 000 0000",
          required: false,
          validation: [
            {
              type: "pattern",
              pattern: "^[0-9]{3}\\s?[0-9]{3}\\s?[0-9]{4}$",
              message: "Enter a 10 digit NHS number",
            },
          ],
        },
      },
    ],
  },
  selection_components: {
    icon: "☑️",
    label: "Selection Components",
    components: [
      {
        type: "select",
        icon: "🔽",
        label: "Select Dropdown",
        description: "Single selection dropdown",
        properties: {
          label: "Select Dropdown",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
      {
        type: "multi_select",
        icon: "⏬",
        label: "Multi Select",
        description: "Multiple selection dropdown",
        properties: {
          label: "Multi Select",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
      {
        type: "checkbox",
        icon: "✅",
        label: "Checkbox",
        description: "Single checkbox input",
        properties: {
          label: "Checkbox",
          required: false,
        },
      },
      {
        type: "radio_group",
        icon: "🔘",
        label: "Radio Group",
        description: "Single selection from multiple options",
        properties: {
          label: "Radio Group",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
    ],
  },
  special_components: {
    icon: "⭐",
    label: "Special Components",
    components: [
      {
        type: "date_picker",
        icon: "📅",
        label: "Date Picker",
        description: "Calendar date selection",
        properties: {
          label: "Date Picker",
          required: false,
        },
      },
      {
        type: "file_upload",
        icon: "📤",
        label: "File Upload",
        description: "File selection and upload",
        properties: {
          label: "File Upload",
          required: false,
        },
      },
      {
        type: "signature",
        icon: "✒️",
        label: "Signature Pad",
        description: "Digital signature capture",
        properties: {
          label: "Signature Pad",
          required: false,
        },
      },
    ],
  },
  layout_components: {
    icon: "📐",
    label: "Layout Components",
    components: [
      {
        type: "horizontal_layout",
        icon: "↔️",
        label: "Horizontal Layout",
        description: "Container for side-by-side components",
        properties: {},
      },
      {
        type: "vertical_layout",
        icon: "↕️",
        label: "Vertical Layout",
        description: "Container for stacked components",
        properties: {},
      },
    ],
  },
  ui_components: {
    icon: "🎨",
    label: "UI Components",
    components: [
      {
        type: "section_divider",
        icon: "➖",
        label: "Section Divider",
        description: "Visual separator between form sections",
        properties: {},
      },
      {
        type: "button",
        icon: "🔘",
        label: "Button",
        description: "Action button for form interactions",
        properties: {
          label: "Button",
          variant: "primary",
        },
      },
      {
        type: "heading",
        icon: "🇭",
        label: "Heading",
        description: "Section heading or title",
        properties: {
          text: "Heading",
          level: "h2",
        },
      },
      {
        type: "card",
        icon: "📇",
        label: "Card Container",
        description: "Visual card container for grouping",
        properties: {},
      },
    ],
  },
};
