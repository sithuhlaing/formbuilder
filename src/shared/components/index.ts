/**
 * Shared Components - Clean Export
 */

export { Button } from './Button';
export { Modal } from './Modal';
export { ActionButton } from './ActionButton';
export { ConfirmDialog } from './ConfirmDialog';
export { DragLayer } from './DragLayer';

// Form Components
export { FormField } from './FormField';
export type { FormFieldProps } from './FormField';

// Input Components
export { 
  TextInput, 
  NumberInput, 
  Textarea, 
  Select, 
  Checkbox, 
  RadioGroup, 
  FileUpload 
} from './Input';

// Specialized Form Components
export {
  SectionDivider,
  SignatureField,
  ButtonPreview,
  HeadingPreview,
  CardPreview,
  DatePicker,
  UnknownComponent
} from './SpecializedFormComponents';

// Simple Component Renderer - Phase 4 Implementation
export { renderSimpleComponent } from '../../components/SimpleRenderer';

// Legacy components have been removed - using simplified architecture
// Current drag-drop functionality in:
// - SimpleCanvas 
// - SimpleDraggableComponent 
// - SimpleComponentPalette