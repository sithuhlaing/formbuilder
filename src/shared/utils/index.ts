
export { generateId } from './generateId';
export { classNames } from './classNames';
// debounce and deepClone removed - were unused

// Type exports for utilities
export type ComponentTypesMap = typeof COMPONENT_TYPES;

// Constants
export const COMPONENT_TYPES = {
  TEXT_INPUT: 'text_input',
  EMAIL_INPUT: 'email_input',
  PASSWORD_INPUT: 'password_input',
  NUMBER_INPUT: 'number_input',
  TEXTAREA: 'textarea',
  RICH_TEXT: 'rich_text',
  SELECT: 'select',
  MULTI_SELECT: 'multi_select',
  CHECKBOX: 'checkbox',
  RADIO_GROUP: 'radio_group',
  DATE_PICKER: 'date_picker',
  FILE_UPLOAD: 'file_upload',
  SECTION_DIVIDER: 'section_divider',
  SIGNATURE: 'signature',
  BUTTON: 'button',
  HEADING: 'heading',
  CARD: 'card',
  HORIZONTAL_LAYOUT: 'horizontal_layout',
  VERTICAL_LAYOUT: 'vertical_layout',
} as const;
