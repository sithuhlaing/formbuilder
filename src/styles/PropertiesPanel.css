/* Properties Panel Styles */
.properties-panel {
  padding: 1rem;
  background-color: #f8f9fa;
  border-left: 1px solid #dee2e6;
  height: 100%;
  overflow-y: auto;
}

.properties-panel h3 {
  margin-top: 0;
  margin-bottom: 1rem;
  padding-bottom: 0.5rem;
  border-bottom: 1px solid #dee2e6;
}

.form-group {
  margin-bottom: 1rem;
}

.form-group label {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: 500;
}

.form-control {
  width: 100%;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  color: #495057;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  border-radius: 0.25rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}

.form-control:focus {
  border-color: #80bdff;
  outline: 0;
  box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25);
}
