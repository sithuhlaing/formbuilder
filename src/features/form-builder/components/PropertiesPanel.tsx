/**
 * Clean Properties Panel - Form Builder Feature
 * Single purpose: Edit selected component properties
 */

import React from 'react';
import type { FormComponentData } from '../../../core/interfaces/ComponentInterfaces';
import { TEST_IDS } from '../../../__tests__/utils/test-utils';
import '../../../styles/PropertiesPanel.css';

interface PropertiesPanelProps {
  selectedComponent: FormComponentData | null;
  onUpdateComponent: (updates: Partial<FormComponentData>) => void;
}

export const PropertiesPanel = ({ 
  selectedComponent, 
  onUpdateComponent 
}: PropertiesPanelProps) => {
  if (!selectedComponent) {
    return (
      <div className="properties-panel" data-testid={TEST_IDS.PROPERTIES_PANEL}>
        <h3>Properties</h3>
        <div className="properties-panel__empty">
          <p>Select a component to edit its properties</p>
        </div>
      </div>
    );
  }

  const handleChange = (field: string, value: unknown) => {
    onUpdateComponent({ [field]: value });
  };


  return (
    <div className="properties-panel" data-testid={TEST_IDS.PROPERTIES_PANEL}>
      <h3>Properties</h3>
      <div className="properties-form">
        {/* Common properties for all components */}
        <div className="form-group">
          <label htmlFor="component-label">Label</label>
          <input
            id="component-label"
            type="text"
            value={selectedComponent.label || ''}
            onChange={(e) => handleChange('label', e.target.value)}
            className="form-control"
          />
        </div>
        
        <div className="form-group">
          <label htmlFor="component-label">Label</label>
          <input
            id="component-label"
            type="text"
            value={selectedComponent.label || ''}
            onChange={(e) => handleChange('label', e.target.value)}
            className="form-control"
          />
        </div>

        <div className="form-group">
          <label htmlFor="component-placeholder">Placeholder</label>
          <input
            id="component-placeholder"
            type="text"
            value={selectedComponent.placeholder || ''}
            onChange={(e) => handleChange('placeholder', e.target.value)}
            className="form-control"
          />
        </div>

        <div className="form-group">
          <label className="checkbox-label">
            <input
              type="checkbox"
              checked={selectedComponent.required || false}
              onChange={(e) => handleChange('required', e.target.checked)}
            />
            Required field
          </label>
        </div>

        {/* Type-specific properties */}
        {selectedComponent.type === 'select' && (
          <div className="form-group">
            <label htmlFor="component-options">Options (one per line)</label>
            <textarea
              id="component-options"
              value={selectedComponent.options?.join('\n') || ''}
              onChange={(e) => handleChange('options', e.target.value.split('\n').filter(Boolean))}
              className="form-control"
              rows={4}
            />
          </div>
        )}

        {selectedComponent.type === 'textarea' && (
          <div className="form-group">
            <label htmlFor="component-rows">Rows</label>
            <input
              id="component-rows"
              type="number"
              min="2"
              max="10"
              value={selectedComponent.rows || 3}
              onChange={(e) => handleChange('rows', parseInt(e.target.value))}
              className="form-control"
            />
          </div>
        )}

        {/* Button Properties */}
        {selectedComponent.type === 'button' && (
          <>
            <div className="form-group">
              <label htmlFor="button-type">Button Type</label>
              <select
                id="button-type"
                value={selectedComponent.buttonType || 'primary'}
                onChange={(e) => handleChange('buttonType', e.target.value)}
                className="form-control"
              >
                <option value="primary">Primary</option>
                <option value="secondary">Secondary</option>
                <option value="success">Success</option>
                <option value="danger">Danger</option>
                <option value="warning">Warning</option>
                <option value="info">Info</option>
                <option value="light">Light</option>
                <option value="dark">Dark</option>
                <option value="link">Link</option>
              </select>
            </div>

            <div className="form-group">
              <label htmlFor="button-size">Size</label>
              <select
                id="button-size"
                value={selectedComponent.size || 'md'}
                onChange={(e) => handleChange('size', e.target.value)}
                className="form-control"
              >
                <option value="sm">Small</option>
                <option value="md">Medium</option>
                <option value="lg">Large</option>
              </select>
            </div>

            <div className="form-group">
              <label htmlFor="button-variant">Variant</label>
              <select
                id="button-variant"
                value={selectedComponent.variant || 'filled'}
                onChange={(e) => handleChange('variant', e.target.value)}
                className="form-control"
              >
                <option value="filled">Filled</option>
                <option value="outline">Outline</option>
                <option value="text">Text</option>
              </select>
            </div>

            <div className="form-group">
              <label className="checkbox-label">
                <input
                  type="checkbox"
                  checked={selectedComponent.fullWidth || false}
                  onChange={(e) => handleChange('fullWidth', e.target.checked)}
                />
                Full Width
              </label>
            </div>

            <div className="form-group">
              <label className="checkbox-label">
                <input
                  type="checkbox"
                  checked={selectedComponent.disabled || false}
                  onChange={(e) => handleChange('disabled', e.target.checked)}
                />
                Disabled
              </label>
            </div>

            <div className="form-group">
              <label className="checkbox-label">
                <input
                  type="checkbox"
                  checked={selectedComponent.isLoading || false}
                  onChange={(e) => handleChange('isLoading', e.target.checked)}
                />
                Loading State
              </label>
            </div>

            <div className="form-group">
              <label htmlFor="button-icon">Icon (SVG or emoji)</label>
              <input
                id="button-icon"
                type="text"
                value={selectedComponent.icon || ''}
                onChange={(e) => handleChange('icon', e.target.value)}
                className="form-control"
                placeholder="Paste SVG or emoji"
              />
            </div>

            <div className="form-group">
              <label>Icon Position</label>
              <div className="radio-group">
                <label className="radio-label">
                  <input
                    type="radio"
                    name="iconPosition"
                    value="left"
                    checked={!selectedComponent.iconPosition || selectedComponent.iconPosition === 'left'}
                    onChange={(e) => handleChange('iconPosition', e.target.value)}
                  />
                  Left
                </label>
                <label className="radio-label">
                  <input
                    type="radio"
                    name="iconPosition"
                    value="right"
                    checked={selectedComponent.iconPosition === 'right'}
                    onChange={(e) => handleChange('iconPosition', e.target.value)}
                  />
                  Right
                </label>
              </div>
            </div>
          </>
        )}

        {selectedComponent.type === 'heading' && (
          <>
            <div className="form-group">
              <label htmlFor="heading-level">Heading Level</label>
              <select
                id="heading-level"
                value={selectedComponent.level || 2}
                onChange={(e) => handleChange('level', parseInt(e.target.value))}
                className="form-control"
              >
                <option value={1}>Heading 1</option>
                <option value={2}>Heading 2</option>
                <option value={3}>Heading 3</option>
                <option value={4}>Heading 4</option>
                <option value={5}>Heading 5</option>
                <option value={6}>Heading 6</option>
              </select>
            </div>

            <div className="form-group">
              <label>Text Alignment</label>
              <div className="radio-group">
                <label className="radio-label">
                  <input
                    type="radio"
                    name="headingAlign"
                    value="left"
                    checked={!selectedComponent.alignment || selectedComponent.alignment === 'left'}
                    onChange={(e) => handleChange('alignment', e.target.value)}
                  />
                  Left
                </label>
                <label className="radio-label">
                  <input
                    type="radio"
                    name="headingAlign"
                    value="center"
                    checked={selectedComponent.alignment === 'center'}
                    onChange={(e) => handleChange('alignment', e.target.value)}
                  />
                  Center
                </label>
                <label className="radio-label">
                  <input
                    type="radio"
                    name="headingAlign"
                    value="right"
                    checked={selectedComponent.alignment === 'right'}
                    onChange={(e) => handleChange('alignment', e.target.value)}
                  />
                  Right
                </label>
              </div>
            </div>

            <div className="form-group">
              <label htmlFor="heading-color">Text Color</label>
              <input
                id="heading-color"
                type="color"
                value={selectedComponent.color || '#000000'}
                onChange={(e) => handleChange('color', e.target.value)}
                className="form-control"
                style={{ height: '38px' }}
              />
            </div>

            <div className="form-group">
              <label>Margins</label>
              <div className="spacing-controls">
                <div className="spacing-control">
                  <label>Top</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.top || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...selectedComponent.margin,
                      top: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Right</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.right || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...selectedComponent.margin,
                      right: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Bottom</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.bottom || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...selectedComponent.margin,
                      bottom: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Left</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.left || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...selectedComponent.margin,
                      left: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
              </div>
            </div>
          </>
        )}

        {/* Card-specific properties */}
        {selectedComponent.type === 'card' && (
          <>
            <div className="form-group">
              <label>Variant</label>
              <select
                value={selectedComponent.variant || 'elevated'}
                onChange={(e) => handleChange('variant', e.target.value)}
                className="form-control"
              >
                <option value="elevated">Elevated</option>
                <option value="outlined">Outlined</option>
                <option value="filled">Filled</option>
              </select>
            </div>

            <div className="form-group">
              <label>Shadow</label>
              <select
                value={selectedComponent.shadow || 'md'}
                onChange={(e) => handleChange('shadow', e.target.value)}
                className="form-control"
              >
                <option value="none">None</option>
                <option value="sm">Small</option>
                <option value="md">Medium</option>
                <option value="lg">Large</option>
                <option value="xl">Extra Large</option>
              </select>
            </div>

            <div className="form-group">
              <label htmlFor="card-padding">Padding</label>
              <input
                id="card-padding"
                type="text"
                value={selectedComponent.padding || '1.25rem'}
                onChange={(e) => handleChange('padding', e.target.value)}
                className="form-control"
                placeholder="e.g., 1rem or 16px"
              />
            </div>

            <div className="form-group">
              <label htmlFor="card-border-color">Border Color</label>
              <input
                id="card-border-color"
                type="color"
                value={selectedComponent.borderColor || '#e2e8f0'}
                onChange={(e) => handleChange('borderColor', e.target.value)}
                className="form-control"
                style={{ height: '38px' }}
              />
            </div>

            <div className="form-group">
              <label htmlFor="card-bg-color">Background Color</label>
              <input
                id="card-bg-color"
                type="color"
                value={selectedComponent.backgroundColor || '#ffffff'}
                onChange={(e) => handleChange('backgroundColor', e.target.value)}
                className="form-control"
                style={{ height: '38px' }}
              />
            </div>

            <div className="form-group">
              <label className="checkbox-label">
                <input
                  type="checkbox"
                  checked={selectedComponent.showHeader !== false}
                  onChange={(e) => handleChange('showHeader', e.target.checked)}
                />
                Show Header
              </label>
            </div>

            {selectedComponent.showHeader !== false && (
              <div className="form-group">
                <label>Header Alignment</label>
                <div className="radio-group">
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="headerAlign"
                      value="left"
                      checked={!selectedComponent.headerAlign || selectedComponent.headerAlign === 'left'}
                      onChange={(e) => handleChange('headerAlign', e.target.value)}
                    />
                    Left
                  </label>
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="headerAlign"
                      value="center"
                      checked={selectedComponent.headerAlign === 'center'}
                      onChange={(e) => handleChange('headerAlign', e.target.value)}
                    />
                    Center
                  </label>
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="headerAlign"
                      value="right"
                      checked={selectedComponent.headerAlign === 'right'}
                      onChange={(e) => handleChange('headerAlign', e.target.value)}
                    />
                    Right
                  </label>
                </div>
              </div>
            )}

            <div className="form-group">
              <label className="checkbox-label">
                <input
                  type="checkbox"
                  checked={selectedComponent.showFooter === true}
                  onChange={(e) => handleChange('showFooter', e.target.checked)}
                />
                Show Footer
              </label>
            </div>

            {selectedComponent.showFooter && (
              <div className="form-group">
                <label>Footer Alignment</label>
                <div className="radio-group">
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="footerAlign"
                      value="left"
                      checked={!selectedComponent.footerAlign || selectedComponent.footerAlign === 'left'}
                      onChange={(e) => handleChange('footerAlign', e.target.value)}
                    />
                    Left
                  </label>
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="footerAlign"
                      value="center"
                      checked={selectedComponent.footerAlign === 'center'}
                      onChange={(e) => handleChange('footerAlign', e.target.value)}
                    />
                    Center
                  </label>
                  <label className="radio-label">
                    <input
                      type="radio"
                      name="footerAlign"
                      value="right"
                      checked={selectedComponent.footerAlign === 'right'}
                      onChange={(e) => handleChange('footerAlign', e.target.value)}
                    />
                    Right
                  </label>
                </div>
              </div>
            )}

            <div className="form-group">
              <label>Margins</label>
              <div className="spacing-controls">
                <div className="spacing-control">
                  <label>Top</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.top || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...(selectedComponent.margin || {}),
                      top: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Right</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.right || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...(selectedComponent.margin || {}),
                      right: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Bottom</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.bottom || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...(selectedComponent.margin || {}),
                      bottom: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
                <div className="spacing-control">
                  <label>Left</label>
                  <input
                    type="text"
                    value={selectedComponent.margin?.left || '0'}
                    onChange={(e) => handleChange('margin', {
                      ...(selectedComponent.margin || {}),
                      left: e.target.value
                    })}
                    className="form-control"
                    placeholder="0"
                  />
                </div>
              </div>
            </div>
          </>
        )}
      </div>
    </div>
  );
};

export default PropertiesPanel;