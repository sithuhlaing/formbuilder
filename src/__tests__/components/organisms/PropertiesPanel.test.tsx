import { vi, describe, it, expect, beforeEach } from 'vitest';
import type { FormComponentData } from '../../../features/form-builder/types';
import { render, screen, fireEvent } from '../../utils/test-utils';
import { PropertiesPanel } from '../../../features/form-builder/components/PropertiesPanel';

describe('PropertiesPanel', () => {
  const mockOnUpdateComponent = vi.fn();
  
  const textInputComponent: FormComponentData = {
    id: 'text-1',
    fieldId: 'text-1',
    type: 'text_input',
    label: 'Name',
    placeholder: 'Enter your name',
    required: false,
    style: {},
    validation: {}
  } as FormComponentData;

  const selectComponent: FormComponentData = {
    id: 'select-1',
    fieldId: 'select-1',
    type: 'select',
    label: 'Country',
    options: [
      { label: 'USA', value: 'usa' },
      { label: 'Canada', value: 'canada' },
      { label: 'UK', value: 'uk' }
    ],
    required: true,
    style: {},
    validation: {}
  } as FormComponentData;

  const textareaComponent: FormComponentData = {
    id: 'textarea-1',
    fieldId: 'textarea-1',
    type: 'textarea',
    label: 'Description',
    placeholder: 'Enter description',
    required: true,
    style: {},
    validation: {}
  } as FormComponentData;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('shows empty state when no component is selected', () => {
    render(<PropertiesPanel selectedComponent={null} onUpdateComponent={mockOnUpdateComponent} />);
    
    expect(screen.getByText('Properties')).toBeInTheDocument();
    expect(screen.getByText('Select a component to edit its properties')).toBeInTheDocument();
  });

  describe('Common Properties', () => {
    it('renders common properties for a text input component', () => {
      render(
        <PropertiesPanel 
          selectedComponent={textInputComponent} 
          onUpdateComponent={mockOnUpdateComponent} 
        />
      );

      // Check common fields
      expect(screen.getByLabelText('Label')).toHaveValue('Name');
      expect(screen.getByLabelText('Placeholder')).toHaveValue('Enter your name');
      expect(screen.getByLabelText('Required field')).not.toBeChecked();
    });

    it('updates label when changed', () => {
      render(
        <PropertiesPanel 
          selectedComponent={textInputComponent} 
          onUpdateComponent={mockOnUpdateComponent} 
        />
      );

      const labelInput = screen.getByLabelText('Label');
      fireEvent.change(labelInput, { target: { value: 'Full Name' } });
      
      expect(mockOnUpdateComponent).toHaveBeenCalledWith({ label: 'Full Name' });
    });

    it('updates placeholder when changed', () => {
      render(
        <PropertiesPanel 
          selectedComponent={textInputComponent} 
          onUpdateComponent={mockOnUpdateComponent} 
        />
      );

      const placeholderInput = screen.getByLabelText('Placeholder');
      fireEvent.change(placeholderInput, { target: { value: 'Type your full name' } });
      
      expect(mockOnUpdateComponent).toHaveBeenCalledWith({ placeholder: 'Type your full name' });
    });

    it('toggles required field', () => {
      render(
        <PropertiesPanel 
          selectedComponent={textInputComponent} 
          onUpdateComponent={mockOnUpdateComponent} 
        />
      );

      const requiredCheckbox = screen.getByLabelText('Required field');
      fireEvent.click(requiredCheckbox);
      
      expect(mockOnUpdateComponent).toHaveBeenCalledWith({ required: true });
    });
  });

  describe('Type-Specific Properties', () => {
    describe('Select Component', () => {
      it('renders options textarea for select component', () => {
        const options = ['USA', 'Canada', 'UK'];
        // Using type assertion to bypass TypeScript error for test data
        const selectComponent = {
          ...textInputComponent,
          type: 'select',
          options: options
        } as unknown as FormComponentData;

        render(<PropertiesPanel selectedComponent={selectComponent} onUpdateComponent={mockOnUpdateComponent} />);

        const optionsTextarea = screen.getByLabelText('Options (one per line)');
        expect(optionsTextarea).toBeInTheDocument();
        expect((optionsTextarea as HTMLTextAreaElement).value).toBe(options.join('\n'));
      });

      it('updates options for select component', () => {
        render(
          <PropertiesPanel 
            selectedComponent={selectComponent} 
            onUpdateComponent={mockOnUpdateComponent} 
          />
        );

        const optionsTextarea = screen.getByLabelText('Options (one per line)');
        fireEvent.change(optionsTextarea, { 
          target: { 
            value: 'USA\nCanada\nUK\nAustralia' 
          } 
        });
        
        expect(mockOnUpdateComponent).toHaveBeenCalledWith({ 
          options: ['USA', 'Canada', 'UK', 'Australia'] 
        });
      });
    });

    describe('Textarea Component', () => {
      it('renders rows input for textarea component', () => {
        const textareaWithRows = { ...textareaComponent, rows: 4 };
        
        render(
          <PropertiesPanel 
            selectedComponent={textareaWithRows} 
            onUpdateComponent={mockOnUpdateComponent} 
          />
        );

        expect(screen.getByLabelText('Rows')).toHaveValue(4);
      });

      it('updates rows for textarea component', () => {
        render(
          <PropertiesPanel 
            selectedComponent={textareaComponent} 
            onUpdateComponent={mockOnUpdateComponent} 
          />
        );

        const rowsInput = screen.getByLabelText('Rows');
        fireEvent.change(rowsInput, { target: { value: '6' } });
        
        expect(mockOnUpdateComponent).toHaveBeenCalledWith({ rows: 6 });
      });
    });
  });

  it('matches snapshot with text input component', () => {
    const { container } = render(
      <PropertiesPanel 
        selectedComponent={textInputComponent} 
        onUpdateComponent={mockOnUpdateComponent} 
      />
    );
    
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with select component', () => {
    const { container } = render(
      <PropertiesPanel 
        selectedComponent={selectComponent} 
        onUpdateComponent={mockOnUpdateComponent} 
      />
    );
    
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with textarea component', () => {
    const textareaWithRows = { ...textareaComponent, rows: 4 };
    
    const { container } = render(
      <PropertiesPanel 
        selectedComponent={textareaWithRows} 
        onUpdateComponent={mockOnUpdateComponent} 
      />
    );
    
    expect(container).toMatchSnapshot();
  });
});
