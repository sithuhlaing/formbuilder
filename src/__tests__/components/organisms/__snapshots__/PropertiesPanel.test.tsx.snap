// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PropertiesPanel > matches snapshot with select component 1`] = `
<div>
  <div
    class="properties-panel"
    data-testid="properties-panel"
  >
    <h3>
      Properties
    </h3>
    <div
      class="properties-form"
    >
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Country"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Country"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-placeholder"
        >
          Placeholder
        </label>
        <input
          class="form-control"
          id="component-placeholder"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="checkbox-label"
        >
          <input
            checked=""
            type="checkbox"
          />
          Required field
        </label>
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-options"
        >
          Options (one per line)
        </label>
        <textarea
          class="form-control"
          id="component-options"
          rows="4"
        >
          [object Object]
[object Object]
[object Object]
        </textarea>
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertiesPanel > matches snapshot with text input component 1`] = `
<div>
  <div
    class="properties-panel"
    data-testid="properties-panel"
  >
    <h3>
      Properties
    </h3>
    <div
      class="properties-form"
    >
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Name"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Name"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-placeholder"
        >
          Placeholder
        </label>
        <input
          class="form-control"
          id="component-placeholder"
          type="text"
          value="Enter your name"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="checkbox-label"
        >
          <input
            type="checkbox"
          />
          Required field
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertiesPanel > matches snapshot with textarea component 1`] = `
<div>
  <div
    class="properties-panel"
    data-testid="properties-panel"
  >
    <h3>
      Properties
    </h3>
    <div
      class="properties-form"
    >
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Description"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-label"
        >
          Label
        </label>
        <input
          class="form-control"
          id="component-label"
          type="text"
          value="Description"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-placeholder"
        >
          Placeholder
        </label>
        <input
          class="form-control"
          id="component-placeholder"
          type="text"
          value="Enter description"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="checkbox-label"
        >
          <input
            checked=""
            type="checkbox"
          />
          Required field
        </label>
      </div>
      <div
        class="form-group"
      >
        <label
          for="component-rows"
        >
          Rows
        </label>
        <input
          class="form-control"
          id="component-rows"
          max="10"
          min="2"
          type="number"
          value="4"
        />
      </div>
    </div>
  </div>
</div>
`;
